src/server.rs

```diff
@@ -1,65 +1,82 @@
 use std::net::SocketAddr;
 use std::ops::DerefMut;
-use std::sync::{Arc, Mutex};
+use std::sync::Arc;
 
 use env_logger::Env;
-use tokio::codec::Decoder;
-use tokio::io;
+use futures::sink::SinkExt;
+use futures::stream::StreamExt;
 use tokio::net::TcpListener;
-use tokio::prelude::*;
 use tokio::runtime;
+use tokio::sync::Mutex;
+use tokio_util::codec::Decoder;
 
 use crate::codec::ABCICodec;
 use crate::messages::abci::*;
 use crate::Application;
 
-/// Creates the TCP server and listens for connections from Tendermint
-pub fn serve<A>(app: A, addr: SocketAddr) -> io::Result<()>
+async fn serve_async<A>(app: A, addr: SocketAddr)
 where
     A: Application + 'static + Send + Sync,
 {
-    env_logger::from_env(Env::default().default_filter_or("info"))
-        .try_init()
-        .ok();
-    let listener = TcpListener::bind(&addr).unwrap();
-    let incoming = listener.incoming();
     let app = Arc::new(Mutex::new(app));
-    let server = incoming
-        .map_err(|err| panic!("Connection failed: {}", err))
-        .for_each(move |socket| {
+    let mut listener = TcpListener::bind(&addr).await.unwrap();
+    while let Some(Ok(socket)) = listener.next().await {
+        let app_instance = app.clone();
+        tokio::spawn(async move {
             info!("Got connection! {:?}", socket);
             let framed = ABCICodec::new().framed(socket);
-            let (_writer, reader) = framed.split();
-            let app_instance = Arc::clone(&app);
-
-            let responses = reader.map(move |request| {
+            let (mut writer, mut reader) = framed.split();
+            let mut mrequest = reader.next().await;
+            while let Some(Ok(ref request)) = mrequest {
                 debug!("Got Request! {:?}", request);
-                respond(&app_instance, &request)
-            });
-
-            let writes = responses.fold(_writer, |writer, response| {
+                let response = respond(&app_instance, request).await;
                 debug!("Return Response! {:?}", response);
-                writer.send(response)
-            });
-            tokio::spawn(writes.then(|x| x.map_err(|_| ()).map(|_| ())))
+                writer.send(response).await.expect("sending back response");
+                mrequest = reader.next().await;
+            }
+            match mrequest {
+                None => {
+                    panic!("connection dropped");
+                }
+                Some(Err(e)) => {
+                    panic!("decoding error: {:?}", e);
```

Comment:
will that stop the server? If yes, is that intended?

Reply:
Yes, it will; it's intended (as it's most likely some unrecoverable error -- e.g. other process than TM connecting to that socket and sending data to it)